File: backend/internal/service/audit.go
package service

import (
	"log"
	"time"

	"gorm.io/gorm"
)

// AuditLog represents a security event.
type AuditLog struct {
	ID        uint      `json:"id" gorm:"primaryKey"`
	UserID    uint      `json:"user_id" gorm:"index"`
	Username  string    `json:"username" gorm:"size:64"`
	Action    string    `json:"action" gorm:"size:32;index"` // login|logout|create|delete|refresh|export
	Resource  string    `json:"resource" gorm:"size:64"`     // subscription|user|webhook
	Detail    string    `json:"detail" gorm:"size:256"`
	IP        string    `json:"ip" gorm:"size:45"`
	Success   bool      `json:"success"`
	CreatedAt time.Time `json:"created_at"`
}

// AuditService handles audit logging.
type AuditService struct {
	db *gorm.DB
}

func NewAuditService(db *gorm.DB) *AuditService {
	return &AuditService{db: db}
}

// Log records an audit event.
func (s *AuditService) Log(userID uint, username, action, resource, detail, ip string, success bool) {
	entry := AuditLog{
		UserID:   userID,
		Username: username,
		Action:   action,
		Resource: resource,
		Detail:   detail,
		IP:       ip,
		Success:  success,
	}

	if err := s.db.Create(&entry).Error; err != nil {
		log.Printf("Audit log failed: %v", err)
	}
}

// List returns audit logs with pagination.
func (s *AuditService) List(page, pageSize int, action string) ([]AuditLog, int64, error) {
	var logs []AuditLog
	var total int64
	q := s.db.Model(&AuditLog{})
	if action != "" {
		q = q.Where("action = ?", action)
	}
	q.Count(&total)
	err := q.Order("id DESC").
		Offset((page - 1) * pageSize).
		Limit(pageSize).
		Find(&logs).Error
	return logs, total, err
}

// Migrate creates the audit_logs table.
func (s *AuditService) Migrate() error {
	return s.db.AutoMigrate(&AuditLog{})
}

File: backend-python/models/audit.py
from datetime import datetime
from sqlalchemy import Column, Integer, String, Text, DateTime
from .database import Base
from ..utils.time import get_current_time


class AuditLog(Base):
    __tablename__ = "audit_logs"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, index=True)
    username = Column(String(64))
    action = Column(String(32))
    resource = Column(String(32))
    detail = Column(Text)
    ip = Column(String(64))
    success = Column(Integer, default=1)
    created_at = Column(DateTime, default=get_current_time, index=True)

    def __repr__(self):
        return f"<AuditLog(id={self.id}, action={self.action}, username={self.username})>"
